### DISCUSSION TITLE
opSEM::grad and opSEM::strongGrad

### DISCUSSION THREAD

**Discussion by yslan** [Nek5000/nekRS#583] [Q&A] [answered]
I'm very much confused about the "strong" gradient in opSEM

Given an scalar field, U, I'd expect the strongGrad computes the derivatives directly and the weak form is the one contested by the test functions. In Nek5000, I think we use this notation.
- strong (PDE) (Nek5000's gradm1)
  $$\nabla U = (\frac{\partial U}{\partial x}, \frac{\partial U}{\partial y}, \frac{\partial U}{\partial z})$$
- weak (PDE) (Nek5000's opgrad or wgradm1)
  (assembled) $$\int v\nabla U = (\int v\frac{\partial U}{\partial x}, \int v\frac{\partial U}{\partial y}, \int v\frac{\partial U}{\partial z})$$

However, the opSEM seems to follow the "strong formulation" and "weak formulation" in discontinuous Garlerkin, so both form are inside an integral.
- opSEM::grad (weak DG) (?)
   $$\int v\nabla U = (\int \frac{\partial v}{\partial x}U, \int \frac{\partial v}{\partial y}U, \int \frac{\partial v}{\partial z}U)$$
- opSEM::strongGrad (strong DG)
   $$\int v\nabla U = (\int v\frac{\partial U}{\partial x}, \int v\frac{\partial U}{\partial y}, \int v\frac{\partial U}{\partial z})$$

We are also forced to compute an gather-scatter averaged derivatives, which is not always what we want.
For example, second order derivatives $$\frac{\partial^2 U}{dx^2}$$ should be compute by the local element-wise operator $$D_x^2 U$$. Doing a gs-avg in-between is not correct and will create extra oscillations from the averaging operator.  That is, I argue, the current opSEM::strongLaplacian is not correct.

In Nek5000's language, opSEM::strongLaplacian currently computes
```
call opgrad(dUdx,dUdy,dUdz,U)   ! weak, assembled
call dssum(dUdx,lx1,ly1,lz1)
call dssum(dUdy,lx1,ly1,lz1)
call dssum(dUdz,lx1,ly1,lz1)
call opcolv(dUdx,dUdy,dUdz,binvm1)

call opdiv(divGradU, dUdx,dUdy, dUdz) ! weak, assembled again
call dssum(divGradU,lx1,ly1,lz1)
call col2(divGradU, binvm1, n)
```

while the right one should be
```
call gradm1(dUdx,dUdy,dUdz,U) 
call opdiv(divGradU, dUdx,dUdy, dUdz)  ! weak
call dssum(divGradU,lx1,ly1,lz1)
call col2(divGradU, binvm1, n)         ! taking average after all derivatives is ok
```

**Comment by stgeke** [comment, marked as answer]
`strongGrad(u)` corresponds to `gradm1(dudx, dudy, dudz, u)` 
Note: nek's `opgrad` computes the strong gradient (like gradm1) but the result is weighted.

`grad(u)` corresponds {`cdtp(dudx, u, rxm, ..., 1)`, `cdtp(dudy, u, rym, ..., 2)`, `cdtp(dudz, u, rzm, ..., 3)`}
Note: nek's `wgradm1` actually computes the weighted strong gradient - the name might be misleading as `w` stands for weighted **not** weak.

In CG the input to any spatial operator has to be in C0. That's why we call QQt in `opSEM::strongLaplacian` after computing the strongGradient. That's missing (hence incorrect) in your example of a strong Laplacian.

In case of the weak Laplacian, the operator is computed directly (like in the Helmholtz operator) and requires only first-order derivatives in a weak sense.
